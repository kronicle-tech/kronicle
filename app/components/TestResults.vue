<template>
  <table v-if="testResults && testResults.length > 0" class="table table-dark">
    <thead>
      <tr>
        <th>Test</th>
        <th>Priority</th>
        <th>Outcome</th>
        <th>Message</th>
      </tr>
    </thead>
    <tbody>
      <tr
        v-for="(testResult, testResultIndex) in sortedTestResults"
        :key="testResultIndex"
      >
        <td class="test-id">
          <TestId :test-id="testResult.testId" />
        </td>
        <td class="test-priority">
          {{ testResult.priority }}
        </td>
        <td
          :class="`test-outcome ${
            testResult.outcome === 'pass'
              ? 'table-primary'
              : testResult.outcome === 'fail'
              ? 'table-danger'
              : 'table-light'
          }`"
        >
          {{ testResult.outcome }}
        </td>
        <td class="test-message">
          <Markdown v-if="testResult.message" :markdown="testResult.message" />
        </td>
      </tr>
    </tbody>
  </table>
</template>

<script lang="ts">
import Vue, { PropType } from 'vue'
import { TestResult } from '~/types/kronicle-service'
import Markdown from '~/components/Markdown.vue'
import TestId from '~/components/TestId.vue'
import { compareTestResults } from '~/src/testResultComparator'

export default Vue.extend({
  components: {
    Markdown,
    TestId,
  },
  props: {
    testResults: {
      type: Array as PropType<TestResult[]>,
      default: undefined,
    },
  },
  computed: {
    sortedTestResults() {
      return [...this.testResults].sort(compareTestResults)
    },
  },
})
</script>

<style scoped>
.test-id {
  width: 30%;
}

.test-priority {
  width: 10%;
}

.test-outcome {
  width: 10%;
}
</style>
