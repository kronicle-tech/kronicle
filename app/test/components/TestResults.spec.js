import { mount } from '@vue/test-utils'
import TestResults from '@/components/TestResults.vue'
import {
  expectClassesInTableRows,
  expectSubTextsInTableRows,
} from '~/test/components/tableUtils'

describe('TestResults', () => {
  let propsData
  let wrapper
  const createWrapper = () => {
    wrapper = mount(TestResults, { propsData })
  }

  function expectTestOutcomeClasses(testOutcomeClasses) {
    expectClassesInTableRows(wrapper, 'td.test-outcome', testOutcomeClasses)
  }

  function expectTestIds(testIds) {
    expectSubTextsInTableRows(wrapper, 'td.test-id a', testIds)
  }

  beforeEach(() => {
    propsData = {}
  })

  afterEach(() => {
    wrapper.destroy()
    wrapper = null
  })

  describe('when testResults prop is not set', () => {
    test('renders nothing', () => {
      createWrapper()
      expect(wrapper.html()).toEqual('')
    })
  })

  describe('when testResults prop is set to an empty array', () => {
    beforeEach(() => {
      propsData.testResults = []
    })

    test('renders nothing', () => {
      createWrapper()
      expect(wrapper.html()).toEqual('')
    })
  })

  describe('when testResults prop is set to one test result', () => {
    beforeEach(() => {
      propsData.testResults = [
        {
          testId: 'test-test-id',
          priority: 'low',
          outcome: 'pass',
          message: 'Test Message 1',
        },
      ]
    })

    test('renders a table showing the test result', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
    })
  })

  describe('when a test result has a `pass` outcome', () => {
    beforeEach(() => {
      propsData.testResults = [
        {
          testId: 'test-test-id',
          priority: 'low',
          outcome: 'pass',
          message: 'Test Message 1',
        },
      ]
    })

    test('renders the outcome table cell with a `primary` background colour', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
      expectTestOutcomeClasses(['test-outcome table-primary'])
    })
  })

  describe('when a test result has a `fail` outcome', () => {
    beforeEach(() => {
      propsData.testResults = [
        {
          testId: 'test-test-id',
          priority: 'low',
          outcome: 'fail',
          message: 'Test Message 1',
        },
      ]
    })

    test('renders the outcome table cell with a `danger` background colour', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
      expectTestOutcomeClasses(['test-outcome table-danger'])
    })
  })

  describe('when a test result has a `not-applicable` outcome', () => {
    beforeEach(() => {
      propsData.testResults = [
        {
          testId: 'test-test-id',
          priority: 'low',
          outcome: 'not-applicable',
          message: 'Test Message 1',
        },
      ]
    })

    test('renders the outcome table cell with a `light` background colour', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
      expectTestOutcomeClasses(['test-outcome table-light'])
    })
  })

  describe('when a test result has a `message` containing Markdown', () => {
    beforeEach(() => {
      propsData.testResults = [
        {
          testId: 'test-test-id',
          priority: 'low',
          outcome: 'pass',
          message: 'Text with **bold** formatting',
        },
      ]
    })

    test('renders the test result with the Markdown converted to HTML', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
      expect(wrapper.find('td.test-message div').html()).toEqual(
        expect.stringContaining('<strong>bold</strong>')
      )
    })
  })

  describe('when testResults prop is set to multiple test results', () => {
    beforeEach(() => {
      propsData.testResults = [
        {
          testId: 'test-test-id-1',
          priority: 'low',
          outcome: 'pass',
          message: 'Test Message 1',
        },
        {
          testId: 'test-test-id-2',
          priority: 'medium',
          outcome: 'fail',
          message: 'Test Message 2',
        },
        {
          testId: 'test-test-id-3',
          priority: 'high',
          outcome: 'not-applicable',
          message: 'Test Message 3',
        },
      ]
    })

    test('renders a table showing the test results', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
      expectTestIds(['test-test-id-2', 'test-test-id-1', 'test-test-id-3'])
    })
  })

  describe('when there are test results that have different outcomes', () => {
    beforeEach(() => {
      propsData.testResults = [
        {
          testId: 'test-test-id-1',
          priority: 'low',
          outcome: 'not-applicable',
          message: 'Test Message 1',
        },
        {
          testId: 'test-test-id-2',
          priority: 'medium',
          outcome: 'pass',
          message: 'Test Message 2',
        },
        {
          testId: 'test-test-id-3',
          priority: 'medium',
          outcome: 'fail',
          message: 'Test Message 3',
        },
      ]
    })

    test('renders the test results sorted by outcome, with `fail` first, then `pass`, then `not-applicable`', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
      expectTestIds(['test-test-id-3', 'test-test-id-2', 'test-test-id-1'])
    })
  })

  describe('when there are two test results with the same outcome', () => {
    beforeEach(() => {
      propsData.testResults = [
        {
          testId: 'test-test-id-2',
          priority: 'low',
          outcome: 'pass',
          message: 'Test Message',
        },
        {
          testId: 'test-test-id-1',
          priority: 'low',
          outcome: 'pass',
          message: 'Test Message',
        },
      ]
    })

    test('renders the test results sorted by test-id', () => {
      createWrapper()
      expect(wrapper.html()).toMatchSnapshot()
      expectTestIds(['test-test-id-1', 'test-test-id-2'])
    })
  })
})
