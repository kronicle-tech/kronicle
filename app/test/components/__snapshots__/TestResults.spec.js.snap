// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestResults when a test result has a \`fail\` outcome renders the outcome table cell with a \`danger\` background colour 1`] = `
"<table class=\\"table table-dark\\">
  <thead>
    <tr>
      <th>Test</th>
      <th>Priority</th>
      <th>Outcome</th>
      <th>Message</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id\\">test-test-id</a></td>
      <td class=\\"test-priority\\">
        low
      </td>
      <td class=\\"test-outcome table-danger\\">
        fail
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message 1</p>
        </div>
      </td>
    </tr>
  </tbody>
</table>"
`;

exports[`TestResults when a test result has a \`message\` containing Markdown renders the test result with the Markdown converted to HTML 1`] = `
"<table class=\\"table table-dark\\">
  <thead>
    <tr>
      <th>Test</th>
      <th>Priority</th>
      <th>Outcome</th>
      <th>Message</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id\\">test-test-id</a></td>
      <td class=\\"test-priority\\">
        low
      </td>
      <td class=\\"test-outcome table-primary\\">
        pass
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Text with <strong>bold</strong> formatting</p>
        </div>
      </td>
    </tr>
  </tbody>
</table>"
`;

exports[`TestResults when a test result has a \`not-applicable\` outcome renders the outcome table cell with a \`light\` background colour 1`] = `
"<table class=\\"table table-dark\\">
  <thead>
    <tr>
      <th>Test</th>
      <th>Priority</th>
      <th>Outcome</th>
      <th>Message</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id\\">test-test-id</a></td>
      <td class=\\"test-priority\\">
        low
      </td>
      <td class=\\"test-outcome table-light\\">
        not-applicable
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message 1</p>
        </div>
      </td>
    </tr>
  </tbody>
</table>"
`;

exports[`TestResults when a test result has a \`pass\` outcome renders the outcome table cell with a \`primary\` background colour 1`] = `
"<table class=\\"table table-dark\\">
  <thead>
    <tr>
      <th>Test</th>
      <th>Priority</th>
      <th>Outcome</th>
      <th>Message</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id\\">test-test-id</a></td>
      <td class=\\"test-priority\\">
        low
      </td>
      <td class=\\"test-outcome table-primary\\">
        pass
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message 1</p>
        </div>
      </td>
    </tr>
  </tbody>
</table>"
`;

exports[`TestResults when testResults prop is set to multiple test results renders a table showing the test results 1`] = `
"<table class=\\"table table-dark\\">
  <thead>
    <tr>
      <th>Test</th>
      <th>Priority</th>
      <th>Outcome</th>
      <th>Message</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id-2\\">test-test-id-2</a></td>
      <td class=\\"test-priority\\">
        medium
      </td>
      <td class=\\"test-outcome table-danger\\">
        fail
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message 2</p>
        </div>
      </td>
    </tr>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id-1\\">test-test-id-1</a></td>
      <td class=\\"test-priority\\">
        low
      </td>
      <td class=\\"test-outcome table-primary\\">
        pass
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message 1</p>
        </div>
      </td>
    </tr>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id-3\\">test-test-id-3</a></td>
      <td class=\\"test-priority\\">
        high
      </td>
      <td class=\\"test-outcome table-light\\">
        not-applicable
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message 3</p>
        </div>
      </td>
    </tr>
  </tbody>
</table>"
`;

exports[`TestResults when testResults prop is set to one test result renders a table showing the test result 1`] = `
"<table class=\\"table table-dark\\">
  <thead>
    <tr>
      <th>Test</th>
      <th>Priority</th>
      <th>Outcome</th>
      <th>Message</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id\\">test-test-id</a></td>
      <td class=\\"test-priority\\">
        low
      </td>
      <td class=\\"test-outcome table-primary\\">
        pass
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message 1</p>
        </div>
      </td>
    </tr>
  </tbody>
</table>"
`;

exports[`TestResults when there are test results that have different outcomes renders the test results sorted by outcome, with \`fail\` first, then \`pass\`, then \`not-applicable\` 1`] = `
"<table class=\\"table table-dark\\">
  <thead>
    <tr>
      <th>Test</th>
      <th>Priority</th>
      <th>Outcome</th>
      <th>Message</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id-3\\">test-test-id-3</a></td>
      <td class=\\"test-priority\\">
        medium
      </td>
      <td class=\\"test-outcome table-danger\\">
        fail
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message 3</p>
        </div>
      </td>
    </tr>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id-2\\">test-test-id-2</a></td>
      <td class=\\"test-priority\\">
        medium
      </td>
      <td class=\\"test-outcome table-primary\\">
        pass
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message 2</p>
        </div>
      </td>
    </tr>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id-1\\">test-test-id-1</a></td>
      <td class=\\"test-priority\\">
        low
      </td>
      <td class=\\"test-outcome table-light\\">
        not-applicable
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message 1</p>
        </div>
      </td>
    </tr>
  </tbody>
</table>"
`;

exports[`TestResults when there are two test results with the same outcome renders the test results sorted by test-id 1`] = `
"<table class=\\"table table-dark\\">
  <thead>
    <tr>
      <th>Test</th>
      <th>Priority</th>
      <th>Outcome</th>
      <th>Message</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id-1\\">test-test-id-1</a></td>
      <td class=\\"test-priority\\">
        low
      </td>
      <td class=\\"test-outcome table-primary\\">
        pass
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message</p>
        </div>
      </td>
    </tr>
    <tr>
      <td class=\\"test-id\\"><a href=\\"/tests/test-test-id-2\\">test-test-id-2</a></td>
      <td class=\\"test-priority\\">
        low
      </td>
      <td class=\\"test-outcome table-primary\\">
        pass
      </td>
      <td class=\\"test-message\\">
        <div>
          <p>Test Message</p>
        </div>
      </td>
    </tr>
  </tbody>
</table>"
`;
